bigflow/dataflow.py

```diff
@@ -1,69 +1,204 @@
 import typing
 import logging
+import uuid
 
 from apache_beam import Pipeline
-from apache_beam.options.pipeline_options import PipelineOptions, GoogleCloudOptions
+from apache_beam.options.pipeline_options import PipelineOptions
+from apache_beam.runners.runner import PipelineResult, PipelineState
 
-from bigflow.commons import public
+from typing import Dict, List, Union, Optional, Tuple
+
+try:
+    from typing import TypedDict, Literal
+except ImportError:
+    from typing_extensions import TypedDict, Literal
+
+from bigflow.commons import public, build_docker_image_tag
 from bigflow.workflow import Job, JobContext
-from bigflow.workflow import DEFAULT_EXECUTION_TIMEOUT_IN_SECONDS, DEFAULT_PIPELINE_LEVEL_EXECUTION_TIMEOUT_SHIFT_IN_SECONDS
+
+import bigflow.build.reflect
+
 
 logger = logging.getLogger(__file__)
 
 
+class PipelineOptionsDict(TypedDict, total=False):
+    """Standard dataflow pipeline options.
+
+    See sources for full list of supported options.
+    https://github.com/apache/beam/blob/master/sdks/python/apache_beam/options/pipeline_options.py
+    """
+
+    runner: Literal['DirectRunner', 'DataflowRunner']
+    experiments: List[str]
+    streaming: bool
+
+    job_name: str
+    labels: List[str]
+
+    temp_location: str
+    staging_location: str
+
+    project: str
+    region: str
+    zone: str
+    worker_region: str
+    worker_zone: str
+
+    autoscaling_algorithm: Literal['THROUGHPUT_BASED', 'NONE']
+    machine_type: str
+    num_workers: int
+    max_workers: int
+
+    network: str
+    subnetwork: str
+    use_public_ips: bool
+    service_account_email: str
+    worker_disk_type: str
+    disk_size_gb: int
+    flexrs_goal: str
+
+
 @public()
 class BeamJob(Job):
 
+    pipeline_level_execution_timeout_shift = 120  # 2 minutes
+
     def __init__(
```

Comment:
Didn't you break the backward compatibility (by, for example, moving pipeline options) ?

Reply:
+1

Reply:
Yes, there are few possible incompatibilities: 

1. `pipeline_options` might be `dict` or instance of `PipelineOptions`
in second case there is conversion `PipelineOptions->dict->PipelineOptions`, which potentially may cause problems.
I've added an [assertion](https://github.com/allegro/bigflow/pull/248/commits/b22e9e9935834aa1d2250bb2eaf8e9162b21095f) for this.

2. 'entry_point' might be 2-arg function when 'entry_point_arguments' is None.
Known difference - when entry_point has varargs (like `lambda pipeline, context, *arsg: ...`). In such case 3rd extra arg `{}` is not passed.

3. Some pipeline options are defaulted (aka 'job_name'), but only when they are not provided by user explicitly.
This should not break anything.